### PR TITLE
Add C# dispatch/invocation interceptors

### PR DESCRIPTION
**Pull request by pepone**
This PR adds dispatch and invocation interceptors for C#.

Invocation interceptors can be register with the communicator, and will be used for all invocations that use the communicator.

Dispatch interceptors can be register with the communicator or with an object adapter, the former are used by all object adapters and the later only with the given object adapter.

### CODE REVIEW

File: csharp/src/Ice/Communicator.cs [hunk 1]
```diff
@@ -1011,6 +1016,20 @@ async Task PerformDisposeAsync()
...
+        public void InterceptDispatch(params DispatchInterceptor[] interceptors) =>
+            _dispatchInterceptors.AddRange(interceptors);
+
+        /// <summary>Add one or more invocation interceptors to the list of invocation interceptors, the invocation
+        /// interceptors registered with the communicator are executed in the order they are registered for each
+        /// invocation done with the communicator.</summary>
+        /// <param name="interceptors">The invocation interceptors to register with the communicator.</param>
+        public void InterceptInvocation(params InvocationInterceptor[] interceptors) =>
+            _invocationInterceptors.AddRange(interceptors);
+
```
Comment (externl) [inline csharp/src/Ice/Communicator.cs:1032]: Is this thread safe?

File: csharp/src/Ice/Invoker.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+using System.Threading.Tasks;
+
+namespace ZeroC.Ice
+{
+    /// <summary>An invoker is a delegate used to send an outgoing request.</summary>
+    /// <param name="target">The proxy for the invocation.</param>
+    /// <param name="request">The outgoing request being send.</param>
```
Comment (externl) [inline csharp/src/Ice/Invoker.cs:10]:
```suggestion
    /// <param name="request">The outgoing request being sent.</param>
```

File: csharp/src/Ice/Invoker.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    /// <returns>The incoming response frame.</returns>
+    public delegate ValueTask<IncomingResponseFrame> Invoker(IObjectPrx target, OutgoingRequestFrame request);
+
+    /// <summary>A invocation interceptor can be registered with the Communicator to intercept proxy invocations.
+    /// </summary>
+    /// <param name="target">The proxy for the invocation.</param>
+    /// <param name="request">The outgoing request being send.</param>
```
Comment (externl) [inline csharp/src/Ice/Invoker.cs:17]:
```suggestion
    /// <summary>An invocation interceptor can be registered with a Communicator to intercept proxy invocations.
    /// </summary>
    /// <param name="target">The proxy for the invocation.</param>
    /// <param name="request">The outgoing request being sent.</param>
```

File: csharp/test/Ice/interceptor/.gitignore [hunk 4]
```diff
@@ -0,0 +1,3 @@
+*.html
+*.xml
+*.xaml
```
Comment (externl) [inline csharp/test/Ice/interceptor/.gitignore:3]: Do we really need this file? I don't see any of these files in the directory.

File: csharp/test/Ice/interceptor/AllTests.cs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                            request.AddBinaryContextEntry(110, 110, (ostr, v) => ostr.WriteInt(v));
+                        }
+                        return next(target, request);
+                    });
+
+                communicator.InterceptInvocation(async (target, request, next) =>
+                    {
+                        TestHelper.Assert(request.Context["interceptor-1"] == "interceptor-1");
+                        request.Context["interceptor-2"] = "interceptor-2";
+                        if (ice2)
```
Comment (externl) [inline csharp/test/Ice/interceptor/AllTests.cs:101]: I understand in the tests we know if it's ice1 or ice2, but that's not true in general or interceptors, you might have a mix. It would be nice to use the interceptor parameters to figure out the protocol.

File: csharp/src/Ice/Communicator.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:1032]: I would add the interceptor registration to the Communicator ctor to avoid this issue.

Comment (pepone) [reply csharp/src/Ice/Communicator.cs:1032]: It is not, we can use a thread safe collection.

Comment (externl) [reply csharp/src/Ice/Communicator.cs:1032]: Locking this list on every invocation doesn't seem great either. Maybe they should be added during communicator consturciton?

File: csharp/test/Ice/interceptor/.gitignore (hunk 4 above)
Comment (pepone) [reply csharp/test/Ice/interceptor/.gitignore:3]: No we don't I copied the test folder from Ice/operations and this come from there, I will remove it

File: csharp/test/Ice/interceptor/AllTests.cs (hunk 5 above)
Comment (pepone) [reply csharp/test/Ice/interceptor/AllTests.cs:101]: right this should be `request.Protocol == Protocol.Ice2`

File: csharp/src/Ice/IObject.cs [hunk 6]
```diff
@@ -115,4 +116,42 @@ protected ValueTask<OutgoingResponseFrame> IceD_ice_idsAsync(IncomingRequestFram
                                                           ostr.WriteSequence(ret, OutputStream.IceWriterFromString)));
         }
     }
+
+    internal static class Servant
+    {
+        internal static ValueTask<OutgoingResponseFrame> DispatchWithInterceptorsAsync(
```
Comment (bernardnormier) [inline csharp/src/Ice/IObject.cs:122]: I don't like this dynamic and recursive building of a new interceptor chain for every single dispatch. I think interceptors / interceptor chain should be much more static - set and frozen when the object adapter is activated.

Then, does it make sense to have:
- a public servant.DispatchAsync which bypasses interceptors
- an internal servant.DispatchWithInterceptorAsync that goes through the interceptor chain

It seems to me the default should be to use the interceptors, so if we have 2 APIs with one internal and the other one public, it should be the other way around.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 7]
```diff
@@ -360,6 +362,9 @@ public T AddWithUUID<T>(string facet, IObject servant, ProxyFactory<T> proxyFact
         public T AddWithUUID<T>(IObject servant, ProxyFactory<T> proxyFactory) where T : class, IObjectPrx =>
             AddWithUUID("", servant, proxyFactory);
 
+        public void Intercept(params DispatchInterceptor[] interceptors) =>
+            Interceptors.AddRange(interceptors);
+
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:371]: I think the interceptors should be specified in Activate.

File: csharp/src/Ice/Communicator.cs (hunk 1 above)
Comment (pepone) [reply csharp/src/Ice/Communicator.cs:1032]: I think will be good that plug-ins can add interceptors, but for this to work they have to call a method on the communicator.

File: csharp/src/Ice/OutgoingResponseFrame.cs [hunk 8]
```diff
@@ -17,26 +16,21 @@ public sealed class OutgoingResponseFrame : OutgoingFrame
...
-                var ostr = new OutputStream(key.Protocol.GetEncoding(), data);
-                ostr.WriteByte((byte)ResultType.Success);
-                _ = ostr.WriteEmptyEncapsulation(key.Encoding);
-                return new OutgoingResponseFrame(key.Protocol,
-                                                 key.Encoding,
-                                                 data,
-                                                 new OutputStream.Position(0, 1),
-                                                 ostr.Tail);
-            });
+        public static OutgoingResponseFrame WithVoidReturnValue(Current current)
```
Comment (bernardnormier) [inline csharp/src/Ice/OutgoingResponseFrame.cs:22]: Why did you remove the caching?

File: csharp/src/Ice/Proxy.cs [hunk 9]
```diff
@@ -343,13 +348,51 @@ public static class Proxy
             return new OutgoingResponseFrame(request, response);
         }
 
+        private static Invoker BuildInvokerChain(
```
Comment (bernardnormier) [inline csharp/src/Ice/Proxy.cs:351]: Same comment as for dispatch.

File: csharp/src/Ice/OutgoingResponseFrame.cs (hunk 8 above)
Comment (pepone) [reply csharp/src/Ice/OutgoingResponseFrame.cs:22]: Because an interceptor can write to the binary context of the cached frame

File: csharp/src/Ice/IPlugin.cs [hunk 10]
```diff
@@ -9,6 +9,8 @@ namespace ZeroC.Ice
     public interface IPlugin : IAsyncDisposable
     {
         /// <summary>Perform any necessary initialization steps.</summary>
-        void Initialize();
+        /// <param name="context">The plug-in initialization context allows to register invocation and dispatch
+        /// interceptors, it can also be use to set the communicator logger.</param>
+        void Initialize(PluginInitializationContext context);
```
Comment (externl) [inline csharp/src/Ice/IPlugin.cs:14]:
```suggestion
        /// <param name="context">The plug-in initialization context enables the registration of invocation and dispatch
        /// interceptors, it can also be use to set the communicator logger.</param>
        void Initialize(PluginInitializationContext context);
```

File: scripts/tests/Ice/interceptor.py [hunk 11]
```diff
@@ -2,4 +2,7 @@
 # Copyright (c) ZeroC, Inc. All rights reserved.
 #
 
-TestSuite(__name__, [ClientTestCase(client=Client(props={"Ice.Warn.Dispatch" : 0}))], libDirs=["interceptortest"])
+if isinstance(Mapping.getByPath(__name__), CSharpMapping):
+    TestSuite(__name__, [ClientServerTestCase(), ClientAMDServerTestCase(), CollocatedTestCase()])
+else:
+    TestSuite(__name__, [ClientTestCase(client=Client(props={"Ice.Warn.Dispatch" : 0}))], libDirs=["interceptortest"])
```
Comment (externl) [inline scripts/tests/Ice/interceptor.py:8]: Should we have a `TODO` to fix this in the future?

Comment (pepone) [reply scripts/tests/Ice/interceptor.py:8]: sure I will add one

File: csharp/src/Ice/Communicator.cs [hunk 12]
```diff
@@ -218,14 +223,18 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
             ILogger? logger = null,
             Instrumentation.ICommunicatorObserver? observer = null,
             TlsClientOptions? tlsClientOptions = null,
-            TlsServerOptions? tlsServerOptions = null)
+            TlsServerOptions? tlsServerOptions = null,
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator.cs:226]: Just like for Clone, we should sort these optional parameters in alphabetical order.

File: csharp/src/Ice/Communicator.cs [hunk 13]
```diff
@@ -1011,6 +1038,14 @@ async Task PerformDisposeAsync()
         public IEndpointFactory? IceFindEndpointFactory(Transport transport) =>
             _transportToEndpointFactory.TryGetValue(transport, out IEndpointFactory? factory) ? factory : null;
 
+        // Add one or more dispatch interceptors, only to use by PluginInitializationContext
+        internal void AddDispatchInterceptor(params DispatchInterceptor[] interceptors) =>
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator.cs:1042]: I am pretty sure we don't need params for these methods since they are not called directly.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 14]
```diff
@@ -132,6 +136,9 @@ public async Task ActivateAsync()
                     throw new InvalidOperationException($"object adapter {Name} already activated");
                 }
 
+                Interceptors.AddRange(interceptors);
```
Comment (bernardnormier) [inline csharp/src/Ice/ObjectAdapter.cs:139]: When you construct a communicator, you can specify dispatch interceptors in the constructor. Then you can add more interceptors _after_ these interceptors using plugins. So it seems more logical to me that interceptors specified in Activate are added after the existing interceptors, not before.

File: csharp/src/Ice/PluginInitializationContext.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+
+namespace ZeroC.Ice
+{
+    public sealed class PluginInitializationContext : IDisposable
+    {
+        private Communicator _communicator;
```
Comment (bernardnormier) [inline csharp/src/Ice/PluginInitializationContext.cs:11]: private readonly...

File: csharp/src/Ice/PluginInitializationContext.cs [hunk 16]
```diff
@@ -0,0 +1,53 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+
+namespace ZeroC.Ice
+{
+    public sealed class PluginInitializationContext : IDisposable
```
Comment (bernardnormier) [inline csharp/src/Ice/PluginInitializationContext.cs:9]: Could we use a ref struct for `PluginInitializationContext` to ensure the plugin can't keep it around? (Then no need for the dispose mechanism).

File: csharp/src/Ice/Communicator.cs (hunk 12 above)
Comment (pepone) [reply csharp/src/Ice/Communicator.cs:226]: Added https://github.com/zeroc-ice/ice/issues/997 to fix in a separate PR

File: csharp/src/Ice/Communicator.cs (hunk 13 above)
Comment (pepone) [reply csharp/src/Ice/Communicator.cs]: Fixed

File: csharp/src/Ice/PluginInitializationContext.cs (hunk 15 above)
Comment (pepone) [reply csharp/src/Ice/PluginInitializationContext.cs]: Fixed

File: csharp/src/Ice/PluginInitializationContext.cs (hunk 16 above)
Comment (pepone) [reply csharp/src/Ice/PluginInitializationContext.cs]: Yes fixed, that is much simpler now
